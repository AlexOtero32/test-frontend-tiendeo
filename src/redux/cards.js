import { v4 as uuid } from 'uuid';

export const CREATE_CARD = 'create_card';
export const EDIT_CARD = 'edit_card';
export const DELETE_CARD = 'delete_card';

const initialState = {
    cards: [1, 2, 3],
};

export default (state = initialState, action) => {
    switch (action.type) {
        case CREATE_CARD:
            const card = {
                ...action.card,
                id: uuid(),
                fecha: new Date(),
            };

            state.cards.push(card);

            return state;
        case EDIT_CARD:
            return state;
        case DELETE_CARD:
            return state;
        default:
            return state;
    }
};
