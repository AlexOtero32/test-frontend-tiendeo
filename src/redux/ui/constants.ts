export enum ActionTypes {
  TOGGLE_MODAL = "toggle_modal",
  SWITCH_THEME = "switch_theme",
}

export type Themes = "light" | "dark" | "blue";
