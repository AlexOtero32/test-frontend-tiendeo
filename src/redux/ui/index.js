export const TOGGLE_MODAL = 'toggle_modal';

const initialState = {
    isModalOpen: false,
};

export function toggleModal() {
    return { type: TOGGLE_MODAL };
}

export default (state = initialState, action) => {
    if (action.type === TOGGLE_MODAL) {
        return { isModalOpen: !state.isModalOpen };
    }

    return state;
};
