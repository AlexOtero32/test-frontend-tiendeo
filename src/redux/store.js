import { createStore, combineReducers } from 'redux';
import cardsReducer from './cards';
import { reducer as formReducer } from 'redux-form';

const reducer = combineReducers({
    cards: cardsReducer,
    form: formReducer,
});

export const store = createStore(reducer);

export default store;
