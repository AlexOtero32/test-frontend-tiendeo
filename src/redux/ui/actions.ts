import { ActionTypes, Themes } from "./constants";

interface ToggleModal {
  type: typeof ActionTypes.TOGGLE_MODAL;
}
interface SwitchTheme {
  type: typeof ActionTypes.SWITCH_THEME;
  theme: Themes;
}

export type UIActions = ToggleModal | SwitchTheme;
