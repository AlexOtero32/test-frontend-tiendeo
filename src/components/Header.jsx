import React from 'react';
import { connect } from 'react-redux';

import Droprown from './Dropdown';
import Modal from './Modal';
import { useOutsideAlerter, isObjectEmpty } from '../lib/helpers';

import store from '../redux/store';
import { toggleModal as toggleModalActionCreator } from '../redux/ui';
import { editCard, createCard } from '../redux/cards/actions';

function create(values) {
    store.dispatch(createCard(values));
    store.dispatch(toggleModalActionCreator());
}

function edit(values) {
    store.dispatch(editCard(values));
    store.dispatch(toggleModalActionCreator());
}

const Header = (props) => {
    const { isModalOpen, toggleModal, editing, dispatch } = props;

    const dropdownRef = React.useRef(null);
    const modalRef = React.useRef(null);

    const [isDroprownOpen, setIsDropdownOpen] = React.useState(false);

    useOutsideAlerter(dropdownRef, setIsDropdownOpen);
    useOutsideAlerter(modalRef, toggleModal);

    return (
        <header className="header">
            <div className="container flex justify-between">
                <div>
                    <button
                        className="button button-blue"
                        onClick={toggleModal}
                    >
                        Nueva tarjeta
                    </button>
                    {isModalOpen && (
                        <div className="overlay">
                            <div ref={modalRef}>
                                <Modal
                                    onSubmit={editing ? edit : create}
                                    onClose={() => dispatch(toggleModal())}
                                />
                            </div>
                        </div>
                    )}
                </div>
                <div ref={dropdownRef}>
                    <button
                        className="button button-border"
                        onClick={() => setIsDropdownOpen(true)}
                    >
                        Ordenar por
                    </button>
                    {isDroprownOpen && <Droprown />}
                </div>
            </div>
        </header>
    );
};

const mapStateToProps = (state) => {
    return {
        isModalOpen: state.ui.isModalOpen,
        editing: !isObjectEmpty(state.cards.editingCard),
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        toggleModal: () => dispatch(toggleModalActionCreator()),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
