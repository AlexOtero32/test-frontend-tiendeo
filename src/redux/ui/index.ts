import { UIActions } from "./actions";
import { ActionTypes, Themes } from "./constants";

export interface UIState {
  isModalOpen: boolean;
  activeTheme: Themes;
}

const initialState: UIState = {
  isModalOpen: false,
  activeTheme: "light",
};

export function toggleModal(): UIActions {
  return { type: ActionTypes.TOGGLE_MODAL };
}

export function switchThemeAction(theme: Themes): UIActions {
  return { type: ActionTypes.SWITCH_THEME, theme };
}

export default (state = initialState, action: UIActions): UIState => {
  switch (action.type) {
    case ActionTypes.TOGGLE_MODAL:
      return Object.assign(state, { isModalOpen: !state.isModalOpen });
    case ActionTypes.SWITCH_THEME:
      return Object.assign(state, { activeTheme: action.theme });
    default:
      return state;
  }
};
