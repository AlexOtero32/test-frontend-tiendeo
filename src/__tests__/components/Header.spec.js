import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Header from "../../components/Header";
import { toggleModal } from "../../redux/ui";

const mockStore = configureStore([]);

describe("Header component", function () {
  let store;

  const initialState = {
    ui: {
      isModalOpen: false,
    },
    cards: { editingCard: {} },
  };

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it("Renderiza correctamente", function () {
    const component = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Botón 'Crear tarjeta' abre el modal", function () {
    store.dispatch = jest.fn();

    const component = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    );

    renderer.act(() => {
      component.root.findAllByType("button")[0].props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledWith(toggleModal());
  });

  it("Botón 'ordenar' abre el menú de ordenación", function () {
    const setState = jest.fn();
    const stateSpy = jest.spyOn(React, "useState");
    stateSpy.mockImplementation(init => [init, setState]);

    const component = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    );

    renderer.act(() => {
      component.root.findAllByType("button")[1].props.onClick();
    });

    expect(setState).toHaveBeenCalledWith(true);
  });
});
