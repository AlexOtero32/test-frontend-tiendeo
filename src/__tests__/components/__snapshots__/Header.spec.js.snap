// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Renderiza correctamente 1`] = `
<header
  className="header"
>
  <div
    className="container flex justify-between"
  >
    <div>
      <button
        className="button button-blue"
        onClick={[Function]}
      >
        Nueva tarjeta
      </button>
    </div>
    <div>
      <button
        className="button button-border"
        onClick={[Function]}
      >
        Ordenar por
      </button>
    </div>
  </div>
</header>
`;
